File: mmoTest/Assets/Script/Data/Proto/WorldMap/WorldMap_CurrRoleUpdateInfoProto.cs
//===================================================
//作    者：边涯  http://www.u3dol.com  QQ群：87481002
//创建时间：2018-02-25 22:40:38
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 客户端发送角色更新信息消息
/// </summary>
public struct WorldMap_CurrRoleUpdateInfoProto : IProto
{
    public ushort ProtoCode { get { return 13013; } }

    public int RoldId; //角色编号
    public string RoleNickName; //角色昵称

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(RoldId);
            ms.WriteUTF8String(RoleNickName);
            return ms.ToArray();
        }
    }

    public static WorldMap_CurrRoleUpdateInfoProto GetProto(byte[] buffer)
    {
        WorldMap_CurrRoleUpdateInfoProto proto = new WorldMap_CurrRoleUpdateInfoProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.RoldId = ms.ReadInt();
            proto.RoleNickName = ms.ReadUTF8String();
        }
        return proto;
    }
}

File: MMORPGServer/Proto/CreateLuaProto/WorldMap_CurrRoleUpdateInfoProto.lua
--客户端发送角色更新信息消息
WorldMap_CurrRoleUpdateInfoProto = { ProtoCode = 13013, RoldId = 0, RoleNickName = "" }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
WorldMap_CurrRoleUpdateInfoProto.__index = WorldMap_CurrRoleUpdateInfoProto;

function WorldMap_CurrRoleUpdateInfoProto.New()
    local self = { }; --初始化self
    setmetatable(self, WorldMap_CurrRoleUpdateInfoProto); --将self的元表设定为Class
    return self;
end


--发送协议
function WorldMap_CurrRoleUpdateInfoProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.RoldId);
    ms:WriteUTF8String(proto.RoleNickName);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function WorldMap_CurrRoleUpdateInfoProto.GetProto(buffer)

    local proto = WorldMap_CurrRoleUpdateInfoProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.RoldId = ms:ReadInt();
    proto.RoleNickName = ms:ReadUTF8String();

    ms:Dispose();
    return proto;
end